### DISCUSSION TITLE
AppleColor High-Resolution RGB Monitor Compatibility

### DISCUSSION THREAD

**Discussion by rdagger** [markadev/AppleII-VGA#21] [Q&A] [answered]
Would your adapter be compatible with one of the early Mac high-res monitors such as the AppleColor High-Resolution RGB Monitor (M1297 or M0401)?  These are the 640x480, 66.7 Hz vertical, 35 kHz horizontal, 69 dpi monitors with the Mac DA-15 connector.

https://support.apple.com/kb/sp418?locale=en_US

**Comment by markadev** [comment]
Hi there! The screen resolution looks compatible but I don't think it could be directly connected due to the different in operating frequencies.  The VGA output is 60 Hz, 31.46875 kHz horizontal, with a pixel frequency of 25.175 MHz so is just slightly different.  I know that it's possible to connect some multisync VGA monitors to Macintoshes but that AppleColor monitor doesn't look like it supports multisync or any other operating frequencies.

Another potential incompatibility could be hsync and vsync signal polarity that the monitor expects.  It's probably documented somewhere but I wasn't able to find it with a few Google searches.  The VGA output is using active low pulses for the hsync and vsync signals.

**Comment by rdagger** [comment]
Thanks for the quick reply!  I figured it was unlikely.  These monitors are plentiful and cheap in my area and are a good aesthetic match for my Apple II.  How difficult would it be to modify the VGA code to match the different rates?
![AppleColor High-Resolution RGB Monitor](https://github.com/markadev/AppleII-VGA/assets/106355/512a1f5b-a53b-4c31-9f32-7dc9efeb3f3d)

Would a later model Apple Multiple Scan monitor work?  It has a VGA mode of 640x480, 60  Hz vertical, 31.77 kHz horizontal:
https://support.apple.com/kb/SP402?locale=en_US

**Reply by markadev** [reply, marked as answer]
That second monitor does look like it would be very compatible given the specs and the info [from this manual](https://www.applerepairmanuals.com/the_manuals_are_in_here/Multiple_Scan_15_Display.pdf).  The repair manual goes into details even on the sync polarity & this monitor apparently supports both 😀 

For the first monitor it's probably possible to modify the current firmware but since the frequency is so far off it could be challenging due to the way that timing is derived throughout all the hardware.  Basically you'd have to:
 1. Change the pixel frequency constant in the code to something close that the monitor is expecting.  Maybe 28 MHz is close enough while also being nice and round.  The rest of the VGA signals are derived from this main frequency so that might be enough to get a basic picture out (assuming the sync signal polarity is correct!)
 2. Update the main Pico's clock speed to be an integer multiple of the pixel frequency chosen.  If it's not then pixels will jitter around due to the non-uniform pixel clock generated by the clock divider in the Pico's PIO engine.  The Pico's clock speed also has to be something that the RP2040's main clock generator can create. I don't know offhand if 140MHz is possible and this will probably need a minor overclocking of the Pico in any case.
 3. If you increased the Pico's main clock speed then the timings in the PIO program that snoops the Apple bus would probably also need to be adjusted to compensate so that it's not looking at the bus signals too early.
 4. You'd also need to keep an eye out for "late" rendered scanlines if the Pico can't keep up rendering with the higher 66.7 Hz frame rate in the higher resolution Apple video modes.  It's probably no issue though if you're running at 140MHz or higher.

**Comment by rdagger** [comment]
I found the pinout for the DB-15 connector if anyone needs it:
![image](https://github.com/markadev/AppleII-VGA/assets/106355/2a62400a-4eea-47e2-92e9-dd094ba58551)
